src/ui/js/Newgame.js

```diff
@@ -572,14 +615,40 @@ Newgame.getSelectOptionList = function(
   return optionlist;
 };
 
+Newgame.getCustomRecipeTd = function(player) {
+  var inputFieldName = player + '_custom_recipe';
+  var currentValue =
+    (player == 'player' ?
+    Newgame.activity.playerCustomRecipe :
+    Newgame.activity.opponentCustomRecipe);
+
+  var customRecipeInput = $('<input>', {
+    'id': inputFieldName,
+    'name': inputFieldName,
+    'type': 'text',
+    'val': currentValue,
+    'style': 'display: none;',
+  });
+
+  var customRecipeTd = $('<td>');
+  customRecipeTd.append($('<span>', {
+    'text': 'Custom Recipe: ',
+    'style': 'display: none;',
+  }));
+  customRecipeTd.append(customRecipeInput);
+
+  return customRecipeTd;
+};
+
 Newgame.getButtonSelectTd = function(player) {
   if (player == 'player') {
     return Newgame.getSelectTd(
       'Your button',
       'player_button',
       Newgame.activity.buttonList.player,
       Newgame.activity.buttonGreyed,
-      Newgame.activity.playerButton);
+      Newgame.activity.playerButton,
+      'Newgame.reactToButtonChange("' + player + '", $(this).val())');
```

Comment:
Drive-by question: without really reading the code, this line looks pretty gross relative to the surrounding text; lots of complicated quoting and stuff.  What's it doing?

Reply:
It's setting up the callback which triggers when a custom button is selected/deselected. The callback selectively shows/hides the relevant input field dependent on whether the custom button is being selected/deselected for the player or the opponent.